### DISCUSSION TITLE
Help understanding bounding box coordinates

### DISCUSSION THREAD

**Discussion by agaeki** [robertknight/ocrs#74] [Q&A]
Hello,

I'm trying to write a program that can recognise text from an image and highlight in the image where the text is, but the bounding boxes returned by engine.detect_words don't seem to have coordinates that make any sense in the image e.g. text that is on opposite sides of the image have similar coordinates, coordinates are sometimes out of range, height and width don't seem to be in pixel units.  Can anyone help me understand how the coordinates of RotatedRect::center() work?

Thanks,
agaeki

**Comment by robertknight** [comment]
> e.g. text that is on opposite sides of the image have similar coordinates, coordinates are sometimes out of range, height and width don't seem to be in pixel units. 

The coordinates returned by eg. `RotatedRect::center` are in pixel coordinates of the input image. You can use the [`bounding_rect`](https://docs.rs/rten-imageproc/latest/rten_imageproc/trait.BoundingRect.html#tymethod.bounding_rect) method to get an axis-aligned bounding rect. There is no guarantee of the orientation of the rect (ie. which way it's "up" vector points). You can use [`orient_towards`](https://docs.rs/rten-imageproc/latest/rten_imageproc/struct.RotatedRect.html#method.orient_towards) to normalize this.

The rotated rects are formed by finding groups of connected text pixels in the input image and then determining the smallest rotated rect which encloses them. If the word is rotated, it is possible that some corners of such a rect may extend outside the boundaries of the input image. You can visualize the binary mask from which the rotated rects are formed by using the `--text-mask` option of the `ocrs` CLI tool:

```
# Generates text-mask.png
ocrs image.png --text-mask
```

You can visualize the word rects using the `--png` option of the `ocrs` CLI tool:

```
ocrs image.jpg --png --output annotated.png
```

The code which draws `annotated.png` is [here](https://github.com/robertknight/ocrs/blob/ea67c9e70adf834e8a8632e50845b51fb38b1bc7/ocrs-cli/src/output.rs#L116).
